pkg/rpcmiddleware/rpcretry/client_interceptors.go

```diff
@@ -0,0 +1,183 @@
+			retryTimeout := callOpts.timeout
+			if callOpts.enableMetadata {
+				if headerTimeout := getHeaderLimiterTimeout(ctx, md, callOpts); headerTimeout > 0 {
+					retryTimeout = headerTimeout
+				}
+			}
+
+			err = waitRetryBackoff(ctx, retryTimeout, attempt)
+			if err != nil {
+				logger.WithError(err).Debug("An unexpected error occurred while in the timeout for the next request retry")
+				return err
+			}
+
+			callCtx := context.WithValue(ctx, attemptMetadataKey, attempt)
+			err = invoker(callCtx, method, req, reply, cc, append(opts, grpc.Header(&md))...)
+			if err == nil {
+				return nil
+			}
+
+			if !isRetriable(err, callOpts) {
+				return err
+			}
+		}
+
+		return err
+	}
+}
+
+// StreamClientInterceptor returns a new streaming client interceptor that retries the execution of external gRPC
+// calls, the retry attempt will only occur if any of the validators define the error as a trigger.
+func StreamClientInterceptor(opts ...Option) grpc.StreamClientInterceptor {
+	callOpts := evaluateClientOpt(opts...)
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		if callOpts.max == 0 {
+			return streamer(ctx, desc, cc, method, opts...)
+		}
+		if desc.ClientStreams {
+			return nil, errInvalidStreamType.New()
+		}
+
+		var err error
+		logger := log.FromContext(ctx)
+
+		for attempt := uint(0); attempt <= callOpts.max; attempt++ {
+			retryTimeout := callOpts.timeout
+			err = waitRetryBackoff(ctx, retryTimeout, attempt)
+			if err != nil {
+				logger.WithError(err).Debug("An unexpected error occurred while in the timeout for the next request retry")
+				return nil, err
+			}
+
+			var clientStream grpc.ClientStream
+			callCtx := context.WithValue(ctx, attemptMetadataKey, attempt)
+			clientStream, err = streamer(callCtx, desc, cc, method, opts...)
+			if err == nil {
+				retriableStream := &retriableStreamClient{
+					ClientStream: clientStream,
+					callOpts:     callOpts,
+					parentCtx:    ctx,
+					streamerCall: func(ctx context.Context) (grpc.ClientStream, error) {
+						return streamer(ctx, desc, cc, method, opts...)
+					},
+				}
+				return retriableStream, nil
+			}
+
+			if !isRetriable(err, callOpts) {
+				return clientStream, err
+			}
+		}
+		return nil, err
+	}
+}
+
+func evaluateClientOpt(opts ...Option) *options {
+	optCopy := &options{}
+	*optCopy = *defaultOptions
+
+	for _, f := range opts {
+		f(optCopy)
+	}
+	return optCopy
+}
+
+func isRetriable(err error, opt *options) bool {
+	if err == nil {
+		return false
+	}
+	for _, check := range opt.validators {
+		if check(err) {
+			return true
+		}
+	}
+	return false
+}
+
+func waitRetryBackoff(ctx context.Context, timeout time.Duration, attempt uint) error {
+	if attempt == 0 {
+		return nil
+	}
+
+	log.FromContext(ctx).WithField("attempt", attempt).WithField("timeout", timeout).Info("Failed request, waiting until next attempt")
```

Comment:
1. Prefer `WithFields` instead of multiple calls to `WithField`
2. Is `Info` appropriate here? Especially considering that the unexpected error above is only a `Debug`?

Reply:
1 - done
2 - I was initially thinking that it would be appropriate to inform the user (outside the debug scope) that a retry attemptis being made, the reason for this was that it might be a good amount of time without any feedback or information. 

Reflecting on it the worst case would be that the user would try the cli command again with `Debug` logs and see that his requests is failing and that it is retrying, so I opted to change the log to `Debug`. The unexpected error log was a mistake and is also fixed